File: LeetCode/7.整数反转.cpp
/*
 * @lc app=leetcode.cn id=7 lang=cpp
 *
 * [7] 整数反转
 */

// @lc code=start
class Solution {
public:
    int reverse(int x) {
        int res = 0, temp;
        while(x != 0){
            temp = x % 10;
            x /= 10;
            // 32位int最大值__INT32_MAX__ + 1 即是 __INT_MIN__
            // __INT_MIN__ = (1 << 31) = 0x80000000
            // __INT_MAX__ = (1 << 31) - 1 = 0x7fffffff
            if(res < (__INT32_MAX__ + 1) / 10 || res > __INT32_MAX__ / 10)
                return 0;
            
            // 多开一个if会多占用资源
            // if(res < (__INT32_MAX__ + 1) / 10) return 0;
            // if(res > __INT32_MAX__ / 10) return 0;

            res = res * 10 + temp;
        }

        return res;
    }
};
// @lc code=end

File: LeetCode/7.整数反转.py
#
# @lc app=leetcode.cn id=7 lang=python3
#
# [7] 整数反转
#

# @lc code=start
class Solution:
    def reverse(self, x: int) -> int:
        # 1. 计算法
        # if -10 < x < 10: return x
        # res, temp = 0, abs(x)
        # overflow = (1 << 31) - 1 if x > 0 else 1 << 31
        # while temp != 0:
        #     res = res * 10 + temp % 10
        #     if res > overflow: return 0
        #     temp //= 10
        # return res if x > 0 else -res

        # 2. string法
        if -10 < x < 10: return x
        str_x = str(x)
        if str_x[0] != '-':
            str_x = str_x[::-1]
            x = int(str_x)
        else:
            str_x = str_x[:0:-1]
            x = int(str_x)
            x = -x
        return x if -2147483648 < x < 2147483647 else 0
# @lc code=end

